src/backend/utils/adt/gp_optimizer_functions.c

```diff
@@ -16,47 +16,78 @@
 #include "funcapi.h"
 #include "utils/builtins.h"
 
-extern Datum EnableXform(PG_FUNCTION_ARGS);
+typedef Datum (*gp_opt_version_func) (void);
+
+/*
+ * A stub function to call if no optimizer function is found.
+ */
+static Datum
+gp_optimizer_function_stub(PG_FUNCTION_ARGS)
+{
+	return CStringGetTextDatum("ORCA is not added to 'shared_preload_libraries'");
+}
+
+/*
+ * Loads optimizer function from a shared library. If library is not presented
+ * or doesn't contain the requested function or any of its dependencies,
+ * returns the stub function.
+ */
+static PGFunction
+gp_optimizer_load_function(char *funcname)
+{
+	volatile PGFunction func = NULL;
+	int32		savedInterruptHoldoffCount = InterruptHoldoffCount;
+	MemoryContext oldcontext = CurrentMemoryContext;
+
+	PG_TRY();
+	{
+		func = load_external_function("$libdir/gp_orca", funcname, false, NULL);
+	}
+	PG_CATCH();
+	{
+		MemoryContextSwitchTo(oldcontext);
+		InterruptHoldoffCount = savedInterruptHoldoffCount;
+		FlushErrorState();
+		func = NULL;
```

Comment:
is this line redundant?

Reply:
I think better to keep it in case one adds any additional code (that can emit an error) into the 'pg_try' block after the var assignment.

Reply:
> in case one adds any additional code (that can emit an error) into the 'pg_try' block after the var assignment.

Why should we consider in this case that the function was not successfully loaded from the library?

Reply:
Ok, let's remove it